### DISCUSSION TITLE
IStreamRequest and IRequestPreProcessor?

### DISCUSSION THREAD

**Discussion by justmara** [LuckyPennySoftware/MediatR#914] [Q&A]
Hi. I've found a bit strange (from my point of view) behaviour of MediatR: for StreamRequests the IRequestPreProcessor is called only on stream iteration begin, not before its creation (is it called PREprocessor?)
```
public record GetDataCommand(long Id): IStreamRequest<byte[]>;

public class GetDataValidator : IRequestPreProcessor<GetDataCommand>
{
    public Task Process(GetDataCommand request, CancellationToken cancellationToken)
    {
        throw new NotImplementedException();
    }
}

var stream = _mediator.CreateStream(new GetDataCommand(1));
await foreach (var buffer in stream) // here the PreProcessor is called
    await output.WriteAsync(buffer);
```

Is that meant to be this way? [Here](https://github.com/jbogard/MediatR/discussions/732#discussioncomment-2406739) it is stated to be `runs before handler`. But streaming case is a bit special because it consists of two calls - stream creation and iteration. So current realisation makes impossible to do command validation behaviour for stream commands with async processing.
btw: IStreamPipelineBehavior acts exactly same - it is invoked only on iteration start, not before.

**Comment by jbogard** [comment]
Stream requests don't use the request pre/post processor stuff. That's only regular requests. Stream requests only have the `IStreamPipelineBehavior` interface. But if that needs to be invoked differently, then there might need to be a completely separate interface for before starting the iteration.
